torch/csrc/jit/codegen/cuda/lower_validation.cpp

```diff
@@ -1037,6 +1037,67 @@ void validateMma(Fusion* fusion) {
   }
 }
 
+namespace {
+
+//! Returns all swizzle ops between the set of iterdomains
+//!  in `from` and `to`.
+std::vector<Expr*> getAllSwizzlesBetween(
+    std::vector<IterDomain*> from,
+    std::vector<IterDomain*> to) {
+  auto all_expr = DependencyCheck::getAllExprsBetween(
+      {from.begin(), from.end()}, {to.begin(), to.end()});
+
+  std::vector<Expr*> all_swizzles;
+
+  std::copy_if(
+      all_expr.begin(),
+      all_expr.end(),
+      std::back_inserter(all_swizzles),
+      [](Expr* expr) {
+        return expr->getExprType().has_value() &&
+            (expr->etype() == ExprType::Swizzle2D);
+      });
+
+  return all_swizzles;
+}
+
+} // namespace
+
+void validateSwizzle(Fusion* fusion) {
+  auto used_vals = fusion->usedMathVals();
+  for (auto tv : ir_utils::filterByType<TensorView>(used_vals)) {
```

Comment:
I'm having trouble parsing this validation pass. It seems to me:
(1) I can't inline a swizzle into anything.
(2) If memory type isn't shared memory I can't inline a swizzle involving thread dims.

I feel like I'm missing something in this validation pass. To me it seems like (2) is redundant to (1). I'm still uncertain in this PR what swizzle support exists and what's validated. I just want to be confident we're validating we currently support cases that come in. I guess I'm comfortable if (2) is redundant, just want to make sure that understanding is correct because I don't understand the block that follows if not.

What exact pattern do we need for MatMul? For transpose we wouldn't try to inline the swizzle in shared memory (if this accidentally happened I think it would be a parallelization/inlining error).

Reply:
I will put this note on the comment in this PR:

There will be 2 types of swizzles, using the same infrastructure, tentatively naming them `persistent swizzle` and `non-persistent swizzle`, (a second candidate pair avoid overloading terms, `loop swizzle (non-persistent)` and `data swizzle(persistent)`). This PR currently only has the `persistent swizzle` in shared memory or register enabled.

Conceptually `persistent swizzle` are the ops that will be "realized" in a memory-like structure, like shared or global memory. It's very similar to the transpose swizzle pattern. e.g.
```python
for i in ...
  for j in ...
   Tswizzled[swizzle(i,j)] = T0[i,j]

// at this point, Tswizzle holds swizzled data pattern. 

for i in ...
  for j in ...
   T1[i,j] = Tswizzled[swizzle(i,j)]
```

In a follow up will be the `non-persistent` swizzle, which will actually be replayed in reference index pass and checked by inline and parallel type validations, and additional checks also needed. In this case, only the loop order will be permuted but no swizzled data pattern should show up on any tensor, e.g.:

``` cpp
for i in ...
  for j in ...
   T1[swizzle(i,j)] = T0[swizzle(i,j)]
```

These are the two patterns needed in MatMul, the `persistent swizzle` in prolog, and `non-persistent swizzle` in block zcurve and register access.

Reply:
On the validation pass in this PR: it is only validating `persistent swizzle`'s, non-parallized or block level parallel into shared memory, so we want to check :
1. It is not inlined, won't support any near term based on the discussion above. This is checking `Tswizzled` in the example above. 

2. If it is block parallel we need to force it in shared memory. I will move this second part in `sync_info` pass.